### PR TITLE
Changed the signal handling to named pipes in the zoo test

### PR DESCRIPTION
**Pull request by raylu-hdf**
For communication between the writer and reader, mainly in the zoo test as a starting point.  The purpose is to make easy porting to Windows.

### CODE REVIEW

File: test/genall5.c [hunk 1]
```diff
@@ -2755,12 +2755,13 @@ tend_zoo(hid_t fid, const char *base_path, struct timespec *lastmsgtime,
     }
 out:
     if (!ok) {
```
Comment (raylu-hdf) [inline test/genall5.c:2756]: The original code set the `lastmsgtime` to 0.  `below_speed_limit` would always return true.  I changed it to use the `lastmsgtime` passed into this function, which is set by the caller.

File: test/testvfdswmr.sh.in [hunk 2]
```diff
@@ -571,24 +571,21 @@ done
 # read and written by VFD SWMR.
 #
 if [ ${do_zoo:-no} = yes ]; then
-	[ -e ./fifo ] && rm -f ./fifo
```
Comment (raylu-hdf) [inline test/testvfdswmr.sh.in:574]: The C program does these two steps now: removing and creating the named pipes.

File: test/testvfdswmr.sh.in [hunk 3]
```diff
@@ -571,24 +571,21 @@ done
...
-	mkfifo -m 0600 ./fifo
 	rm -f ./shared_tick_num
 	echo launch vfd_swmr_zoo_writer
-	STDIN_PATH="./fifo" catch_out_err_and_rc vfd_swmr_zoo_writer \
-	    ../vfd_swmr_zoo_writer -m 1000 -q &
+	catch_out_err_and_rc vfd_swmr_zoo_writer \
+	    ../vfd_swmr_zoo_writer -l 5 -m 1000 -q &
```
Comment (raylu-hdf) [inline test/testvfdswmr.sh.in]: `-l 5` tells the writer that it has 5 seconds to finish the creation or deletion of objects. `-l 10` tells the reader that it has 10 seconds to validate the creation or deletion of objects.

File: test/testvfdswmr.sh.in [hunk 4]
```diff
@@ -571,24 +571,21 @@ done
...
+	catch_out_err_and_rc vfd_swmr_zoo_reader \
+	    ../vfd_swmr_zoo_reader -l 10 -q -W &
 	pid_reader=$!
 
 	# Wait for the reader to finish before signalling the
 	# writer to quit: the writer holds the file open so that the
 	# reader will find the shadow file when it opens
 	# the .h5 file.
 	wait $pid_reader
-	kill -USR1 $(cat vfd_swmr_zoo_writer.pid)
```
Comment (raylu-hdf) [inline test/testvfdswmr.sh.in:591]: I don't know why this is necessary.  The test works fine without it.

File: test/vfd_swmr_zoo_writer.c [hunk 5]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
+
         if (!create_zoo(fid, ".", &lastmsgtime, config))
             errx(EXIT_FAILURE, "create_zoo didn't pass self-check");
 
-        /* Avoid deadlock: flush the file before waiting for the reader's
-         * message.
-         */
-        if (H5Fflush(fid, H5F_SCOPE_GLOBAL) < 0)
```
Comment (raylu-hdf) [inline test/vfd_swmr_zoo_writer.c:352]: The test hangs with this function call.  I'm not sure why.

File: test/vfd_swmr_common.c [hunk 6]
```diff
@@ -68,6 +68,21 @@ below_speed_limit(struct timespec *last, const struct timespec *ival)
     return result;
 }
 
+/* Sleep for `tenths` tenths of a second.
+ *
+ * This routine may quietly perform a too-short sleep if an error occurs
+ * in nanosleep(2).
+ */
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_common.c]: Don't quite understand why this new decisleep() is not using H5_nanosleep() in the library?

File: test/vfd_swmr_zoo_writer.c [hunk 7]
```diff
@@ -108,6 +111,9 @@ vfd_swmr_writer_may_increase_tick_to(uint64_t new_tick, bool wait_for_reader)
     dbgf(3, "%s: enter\n", __func__);
 
     if (fd == -1) {
+        if (access("./shared_tick_num", F_OK ) < 0)
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Use HDaccess().

File: test/vfd_swmr_zoo_writer.c [hunk 8]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
 
-        if (step != 'b')
-            errx(EXIT_FAILURE, "expected 'b' read '%c'", step);
+        /* Wait until the reader finishes validating zoo creation */
+        if (HDread(fd_reader_to_writer, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "read failed");
+        if (notify != 3)
+            errx(EXIT_FAILURE, "expected 2 but read %d", notify);
+
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) == -1)
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Use HDclock_gettime()

File: test/vfd_swmr_zoo_writer.c [hunk 9]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
+        dbgf(2, "To reproduce, set seed (%s) to %u.\n", seedvar, seed);
+
+        /* Writer tells reader to start */
+        notify = 1;
+        if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "write failed");
 
         ostate = initstate(seed, vector, _arraycount(vector));
 
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Use HDclock_gettime().

File: test/vfd_swmr_zoo_writer.c [hunk 10]
```diff
@@ -259,6 +270,18 @@ main(int argc, char **argv)
         case 'e':
             print_estack = true;
             break;
+        case 'l':
+            /* Expected maximal time for reader's validation of zoo creation or deletion */
+            errno = 0;
+            tmpl = strtoul(optarg, &end, 0);
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Use HDstrtoul().

File: test/vfd_swmr_zoo_writer.c [hunk 11]
```diff
@@ -33,6 +33,8 @@
 #define _arraycount(_a) (sizeof(_a)/sizeof(_a[0]))
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Similar definition of __arraycount exists in vfd_swmr.c.  Is it possible to put it in a common header or file so both can use access it?

File: test/vfd_swmr_zoo_writer.c [hunk 12]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
+
         if (!create_zoo(fid, ".", &lastmsgtime, config))
             errx(EXIT_FAILURE, "create_zoo didn't pass self-check");
 
-        /* Avoid deadlock: flush the file before waiting for the reader's
-         * message.
-         */
-        if (H5Fflush(fid, H5F_SCOPE_GLOBAL) < 0)
-            errx(EXIT_FAILURE, "%s: H5Fflush failed", __func__);
+        /* Notify the reader of finishing zoo creation */
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c]: Write description/comments about how pipes are used in coordinating between the writer and reader

File: test/testvfdswmr.sh.in (hunk 4 above)
Comment (gnuoyd) [reply test/testvfdswmr.sh.in:591]: You deleted the part of the writer that waits for the signal.

File: test/genall5.c (hunk 1 above)
Comment (gnuoyd) [reply test/genall5.c:2756]: I think that you have misunderstood `below_speed_limit`. It's a rate-limiting utility. Let me know how I can clarify the comment on `below_speed_limit`.

File: test/genall5.c [hunk 13]
```diff
@@ -2755,12 +2755,13 @@ tend_zoo(hid_t fid, const char *base_path, struct timespec *lastmsgtime,
...
-        if (below_speed_limit(lastmsgtime, &config.msgival)) {
-            last_failure_mssg = failure_mssg;
-            warnx("%s: %s", __func__, failure_mssg);
+        /* Only the zoo test for VFD SWMR does this step, making sure it doesn't take too long.
+         * other tests sets config.msgival to 0 */
+        if (strcmp(failure_mssg, last_failure_mssg) != 0 && ((config.msgival.tv_sec || config.msgival.tv_nsec))) {
+            if (below_speed_limit(lastmsgtime, &config.msgival)) {
+                last_failure_mssg = failure_mssg;
+                warnx("%s: %s", __func__, failure_mssg);
+            }
```
Comment (gnuoyd) [inline test/genall5.c:2764]: I don't understand the importance of this change.

The VFD SWMR reader will produce a deluge of identical error messages while it waits for the content it expects to appear in the HDF5 file that it shares with the writer, unless the error messages are rate-limited. The non-VFD SWMR "zoo" tests are not expected to ever trigger rate-limiting, so limiting the output rate, anyway, does not affect the error output in any significant way.

File: test/testvfdswmr.sh.in (hunk 3 above)
Comment (gnuoyd) [reply test/testvfdswmr.sh.in]: Careful setting tight time limits, slow or overburdened machines will run right over the limits and produce time-wasting spurious failures. A time limit one minute or longer is probably more appropriate.

File: test/vfd_swmr_zoo_writer.c [hunk 14]
```diff
@@ -108,6 +111,9 @@ vfd_swmr_writer_may_increase_tick_to(uint64_t new_tick, bool wait_for_reader)
     dbgf(3, "%s: enter\n", __func__);
 
     if (fd == -1) {
+        if (access("./shared_tick_num", F_OK ) < 0)
+            return true;
+
```
Comment (gnuoyd) [inline test/vfd_swmr_zoo_writer.c:177]: I think that if you need an `access(2)` call, it deserves an explanatory comment. An access check is inherent in the next statement, the `open(2)` call.

File: test/vfd_swmr_zoo_writer.c [hunk 15]
```diff
@@ -324,14 +347,26 @@ main(int argc, char **argv)
...
 
-    if (wait_for_signal)
-        block_signals(&oldsigs);
+    /* Writer creates two named pipes(FIFO) to coordinate two-way communication */
+    if (writer) {
+        if (HDmkfifo(fifo_writer_to_reader, 0666) < 0)
+            errx(EXIT_FAILURE, "HDmkfifo");
+
+        if (HDmkfifo(fifo_reader_to_writer, 0666) < 0)
+            errx(EXIT_FAILURE, "HDmkfifo");
```
Comment (gnuoyd) [inline test/vfd_swmr_zoo_writer.c]: Maybe use 0600 here per the "principle of least authority" ?

File: test/vfd_swmr_zoo_writer.c [hunk 16]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
+        /* Writer tells reader to start */
+        notify = 1;
+        if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "write failed");
 
         ostate = initstate(seed, vector, _arraycount(vector));
 
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
+
```
Comment (gnuoyd) [inline test/vfd_swmr_zoo_writer.c]: What are these`lastmsgtime` re-initializations for?

File: test/vfd_swmr_zoo_writer.c [hunk 17]
```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
...
+        last.tv_nsec = lastmsgtime.tv_nsec;
+
         while (!validate_zoo(fid, ".", &lastmsgtime, config))
             ;
 
-        if (write(STDOUT_FILENO, &step, sizeof(step)) == -1)
-            err(EXIT_FAILURE, "write");
+        /* Make sure zoo validation doesn't take longer than the expected time */
+        if (below_speed_limit(&last, &ival))
+            warnx("validate_zoo took too long to finish");
```
Comment (gnuoyd) [inline test/vfd_swmr_zoo_writer.c]: This isn't really how `below_speed_limit` is meant to be used. :-) You need a function like [`timespeccmp`](https://man.netbsd.org/timespeccmp).

**Review by gnuoyd — CHANGES_REQUESTED** [review]

File: test/genall5.c (hunk 1 above)
Comment (raylu-hdf) [reply test/genall5.c:2756]: This function is no longer used by the zoo test after refactoring the code.  So I will close this comment.

File: test/testvfdswmr.sh.in (hunk 3 above)
Comment (raylu-hdf) [reply test/testvfdswmr.sh.in]: `-l` is the maximal interval (between the writer's finishing zoo creation or deletion and reader's finishing validation) in terms of tick now.

File: test/vfd_swmr_zoo_writer.c (hunk 14 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c:177]: When the file doesn't exist yet, `open` will spill out an error message.

File: test/vfd_swmr_zoo_writer.c (hunk 15 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: You're right.  I fixed it.

File: test/vfd_swmr_zoo_writer.c (hunk 17 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: I think a new function will be similar to `below_speed_limit `

File: test/vfd_swmr_common.c (hunk 6 above)
Comment (raylu-hdf) [reply test/vfd_swmr_common.c]: I fixed it.

File: test/vfd_swmr_zoo_writer.c (hunk 8 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: Fixed.

File: test/vfd_swmr_zoo_writer.c (hunk 9 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: Fixed.

File: test/vfd_swmr_zoo_writer.c (hunk 10 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: Changed.

File: test/vfd_swmr_zoo_writer.c (hunk 11 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: It's not used in the zoo test.  I'll take it out.

File: test/vfd_swmr_zoo_writer.c (hunk 12 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: I'll improve the comments.

**Comment by raylu-hdf** [comment]
This PR is ready for another round of review, hopefully the final one.

File: test/vfd_swmr_zoo_writer.c [hunk 18]
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
...
+    fprintf(stderr,   "  -a: run all tests, including variable-length data\n");
+    fprintf(stderr,   "  -e: print error stacks\n");
+    fprintf(stderr,   "  -l secs: maximal seconds for reader to validate the zoo\n");
+    fprintf(stderr,   "  -p: do not use named pipes\n");
+    fprintf(stderr,   "  -q: be quiet: few/no progress messages\n");
+    fprintf(stderr,   "  -v: be verbose: most progress messages\n");
     exit(EXIT_FAILURE);
 }
 
-bool
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c:100]: Should we also remove the corresponding library instrumentation function of this routine?

File: test/vfd_swmr_zoo_writer.c [hunk 19]
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
...
-    }
-    if (wait_for_reader && !do_try)
-        errx(EXIT_FAILURE, "%s: timed out waiting for reader", __func__);
-
-    tick_stats->writer_aborted_increase++;
-
-    return false;
-}
-
-void
```
Comment (vchoi-hdfgroup) [inline test/vfd_swmr_zoo_writer.c:163]: Should we also remove the corresponding library instrumentation function of this routine?

Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c:163]: That's a good question.  I'll ask John about it in tomorrow's meeting.

File: test/vfd_swmr_zoo_writer.c [hunk 20]
```diff
@@ -74,335 +57,437 @@ print_cache_hits(H5C_t H5_ATTR_UNUSED *cache)
...
-    fprintf(stderr, "  -a: run all tests, including variable-length data\n");
-    fprintf(stderr, "  -e: print error stacks\n");
-    fprintf(stderr, "  -m ms: maximum `ms` milliseconds pause between\n");
-    fprintf(stderr, "         each create/delete step\n");
-    fprintf(stderr, "  -q: be quiet: few/no progress messages\n");
-    fprintf(stderr, "  -v: be verbose: most progress messages\n");
+    fprintf(stderr,   "  -S: do not use VFD SWMR\n");
+    fprintf(stderr,   "  -a: run all tests, including variable-length data\n");
+    fprintf(stderr,   "  -e: print error stacks\n");
+    fprintf(stderr,   "  -l secs: maximal seconds for reader to validate the zoo\n");
```
Comment (kyang2014) [inline test/vfd_swmr_zoo_writer.c]: This "-l secs: maximal seconds for reader to validate the zoo\n" seems not accurate. From the code it seems that it is "maximal number of ticks for reader to validate the zoo\n"

File: test/testvfdswmr.sh.in [hunk 21]
```diff
@@ -571,24 +571,21 @@ done
...
-	STDIN_PATH="./fifo" catch_out_err_and_rc vfd_swmr_zoo_writer \
-	    ../vfd_swmr_zoo_writer -m 1000 -q &
+	catch_out_err_and_rc vfd_swmr_zoo_writer \
+	    ../vfd_swmr_zoo_writer -q &
 	pid_writer=$!
 
-	STDOUT_PATH="./fifo" catch_out_err_and_rc vfd_swmr_zoo_reader \
-	    ../vfd_swmr_zoo_reader -q -W &
+	catch_out_err_and_rc vfd_swmr_zoo_reader \
+	    ../vfd_swmr_zoo_reader -l 4 -q &
```
Comment (kyang2014) [inline test/testvfdswmr.sh.in:583]: Better to add a comment "the expected maximal number of ticks for reader to validate"

**Review by kyang2014 — COMMENTED** [review]
The main() was long before. With the "named pipe" added, the main() function becomes extremely long(more than 400 lines). Is it possible to divide into a few subroutines?

**Comment by gnuoyd** [comment]
The changes in the PR appear to substantially modify the behavior of the test without providing a clear rationale. I am concerned that with the new behavior, there are very few opportunities for the reader and writer to concurrently access the HDF5 file and the metadata file. That is, it's a weaker test than before.

A couple of questions:

What end result are the changes supposed to bring about?

Is there a smaller set of changes that produces the same end result?

**Comment by raylu-hdf** [comment]
> The changes in the PR appear to substantially modify the behavior of the test without providing a clear rationale. I am concerned that with the new behavior, there are very few opportunities for the reader and writer to concurrently access the HDF5 file and the metadata file. That is, it's a weaker test than before.
> 
> A couple of questions:
> 
> What end result are the changes supposed to bring about?
> 
> Is there a smaller set of changes that produces the same end result?

The team discussed the instrumentation part in the zoo test and the driver.  The consensus was that it would be difficult to make it portable, especially on Windows.  Removing it will also make the test easier to maintain.  The concern for the synchronization between the writer and reader is mainly on the time difference between the writer's finishing zoo and the reader's finishing validation.

**Comment by raylu-hdf** [comment]
> The main() was long before. With the "named pipe" added, the main() function becomes extremely long(more than 400 lines). Is it possible to divide into a few subroutines?

I did my best to refactor it and reduced the length of the main function.

File: test/testvfdswmr.sh.in (hunk 21 above)
Comment (raylu-hdf) [reply test/testvfdswmr.sh.in:583]: I did it.

File: test/vfd_swmr_zoo_writer.c (hunk 20 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: I corrected it.

File: test/vfd_swmr_zoo_writer.c (hunk 19 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c:163]: Dana will get rid of them.

File: test/vfd_swmr_zoo_writer.c (hunk 18 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c:100]: Dana will handle it.

File: test/vfd_swmr_zoo_writer.c (hunk 7 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: It's no longer used.

File: test/genall5.c (hunk 13 above)
Comment (raylu-hdf) [reply test/genall5.c:2764]: We talked about it in our meeting.  It should be ok now.

File: test/vfd_swmr_zoo_writer.c (hunk 16 above)
Comment (raylu-hdf) [reply test/vfd_swmr_zoo_writer.c]: We talked about it in our meeting.

**Comment by raylu-hdf** [comment]
I believe I have addressed all of your concerns and comments.  Is this PR good to merge?
@kyang2014 @vchoi-hdfgroup

**Comment by kyang2014** [comment]
The change looks good to me.

**Review by kyang2014 — APPROVED** [review]

**Comment by vchoi-hdfgroup** [comment]
Minor suggestions: 
(1) Move the closing of named pipes to a subroutine: e.g. close_open_named_pipes()
(2) [If (use_named_pipe && create_open_named_pipes() < 0).....]  may simplify the subroutine: create_open_named_pipes()

**Review by vchoi-hdfgroup — APPROVED** [review]

**Comment by raylu-hdf** [comment]
> Minor suggestions:
> (1) Move the closing of named pipes to a subroutine: e.g. close_open_named_pipes()
> (2) [If (use_named_pipe && create_open_named_pipes() < 0).....] may simplify the subroutine: create_open_named_pipes()

I fixed them.

**Pull request merged**